### PR TITLE
Fixes #2592 - Fetches a new milestone.json if needed

### LINKED ISSUE #2592: wrong milestone.json file gives an unhelpful message
Edited by @karlcow (2018-09-25)

See my [comment](https://github.com/webcompat/webcompat.com/issues/2592#issuecomment-424170192) on how to improve this issue. I edited the title too.

Original report by @Obtin 
---
## Steps to reproduce
```
git clone git@github.com:<username>/webcompat.com.git
$ cd webcompat.com.git
$ virtualenv env
$ source env/bin/activate
$ pip2 install -r config/requirements.txt
$ npm install
$ npm run build
$ cp config/secrets.py.example config/secrets.py
```
After which  added the outh token and client id and secret key as per the documentation. After that,
```
$ npm start
```
This is what I got
```
sh: 1: source: not found
Writing logs to: /home/tina/webcompat.com/tmp
Statuses Initialization…
A milestone is missing or has been added: [u'worksforme', u'non-compat', u'needscontact', u'invalid', u'needstriage', u'duplicate', u'needsdiagnosis', u'sitewait', u'wontfix', u'fixed', u'incomplete', u'contactready']
Milestones are not initialized. Check logs.
```

But I checked there are no extra and the test-repo has all these milestones

### PR DESCRIPTION
**Pull request by marimeireles**
This PR fixes issue #2592

To reproduce the error I've followed these steps:

```
git clone git@github.com:<username>/webcompat.com.git
$ cd webcompat.com.git
$ virtualenv env
$ source env/bin/activate
$ pip2 install -r config/requirements.txt
$ npm install
$ npm run build
$ cp config/secrets.py.example config/secrets.py
```

Then, I created a milestone with a typo in my test repo, like `non-compat--`.

If this happens the user will receive the following error message:

```
Writing logs to: /Users/mariana/Development/outreachy/marimeireles/webcompat.com/tmp
Statuses Initialization…
A milestone is missing or has been added: [u'non-compat', u'non-compat--']
Check the milestones names on your Github repository and try again.
This error was probably caused by a typo.
Your milestones.json was erased and a backup copy was created at /Users/mariana/Development/outreachy/marimeireles/webcompat.com/data/milestones_backup.json.
```

When the user types the correct milestone in their repository a new milestones.json file will be downloaded and webcompat will run normally.

This solution is working but I'm not sure if it's the best way to implement it. Please @karlcow let me know if I can improve it!

---

- [x] I have read the [Pull Request + Code Style Guidelines](https://github.com/webcompat/webcompat.com/blob/master/docs/pr-coding-guidelines.md) thoroughly.

### CODE REVIEW

File: config/__init__.py [hunk 1]
```diff
@@ -93,6 +104,7 @@ def update_status_config(milestones_content):
     if not set(status_names).issubset(milestone_names):
         missing = set(status_names).symmetric_difference(set(milestone_names))
         print(MILESTONE_UNMATCHING.format(names=list(missing)))
+        milestones_content = MILESTONE_UNMATCHING
```
Comment (karlcow) [inline config/__init__.py]: This `milestones_content` value is scoped only to `update_status_config`, which means it has no meaning outside of this function. The important part which needs to be checked is the return value of the function. In this case: `None`.

Also note that we are only testing `not set(status_names).issubset(milestone_names)` in that `if`.

File: config/__init__.py [hunk 2]
```diff
@@ -64,6 +70,11 @@ def initialize_status():
             return False
     # milestones_content exists
     updated_statuses = update_status_config(milestones_content)
+    if updated_statuses == MILESTONE_UNMATCHING:
```
[the remark this answers is not in the document]
Comment (karlcow) [reply config/__init__.py]: That said I'm surprised you seem to get the right behavior according to your test.

💭🧐

ah understood. So indeed the function returns `None`. So it runs the code but for the wrong reason.
`None` by default, this triggers a false positive in the code.

**Comment by marimeireles** [comment]
I've commited the wrong version of the code! :(
I wrote that code and then I realized it would crash and I wrote another one.
I'm sorry. I didn't realize it because, as you said, it's returning a false positive.
I'm really sorry, I should have double checked the content of the commit. 
I commited the right version now.

>Also please rebase once you are done. You can make a separate commit for tests (first), then a commit for the function/code.

I did a rebase. As you can see I had some problems with git (grr) but I think it's correct now. The only commit that shows in the branch history is `deec3de`.

>To test this correctly, you need to add tests to [...]

I do need help, thanks!
Doesn't this test already covers the code snippet I just wrote? Because that's what I'm doing, checking if the milestones were typed correctly.
```
        # Unknown milestone added to the project
        milestones_json = json_data('milestones_content_plus.json')
        actual = update_status_config(milestones_json)
        self.assertEqual(actual, expected)
```
If it doesn't could you be more specific on what I should test? I couldn't think of anything else.

Also, do you think the solution I've just commited is okay? I'm not sure if I understood what you meant now that I'm seeing these tests.

Thanks for all the help!

**Comment by miketaylr** [comment]
>  I'm really sorry, I should have double checked the content of the commit.

(no need to apologize, it happens to all of us. :))

**Comment by karlcow** [comment]
@marimeireles aha As @miketaylr said, we do plenty of mistakes, it's part of learning. We should not, but we do. The server went down a couple of times two weeks ago, because @miketaylr and I did some mistakes. 

And the review process for code is exactly for this.

**Comment by karlcow** [comment]
I'll deploy on next monday to avoid Friday-deploy ;)

**Comment by marimeireles** [comment]
okay!
Thanks for the nice comments :)

**Pull request merged**